manager/manager.go

```diff
@@ -0,0 +1,359 @@
+
+					if m.isLeader {
+						m.isLeader = false
+					}
+
+				} else {
+					m.isLeader = evt.Status == leader.Leader
+				}
+				next := m.isLeader
+
+				m.lock.Unlock()
+
+				if current != next {
+					notify <- next
+				}
+
+			}
+		}
+
+	}()
+
+	return m.running, nil
+}
+
+// Stop stops the manager
+func (m *manager) Stop() {
+	if m.stop == nil {
+		return
+	}
+	m.leader.Stop()
+	close(m.stop)
+}
+
+func (m *manager) onAssumeLeadership() error {
+	log.Infoln("Assuming leadership")
+
+	// load the config
+	config := GlobalSpec{}
+
+	// load the latest version -- assumption here is that it's been persisted already.
+	err := m.snapshot.Load(&config)
+	if err != nil {
+		log.Warningln("Error loading config", err)
+		return err
+	}
+
+	log.Infoln("Loaded snapshot. err=", err)
+	if err != nil {
+		return err
+	}
+
+	m.lock.Lock()
+	m.currentConfig = config // make a copy
+	m.lock.Unlock()
+
+	return m.doWatchGroups(config)
+}
+
+func (m *manager) onLostLeadership() error {
+	log.Infoln("Lost leadership")
+
+	// Unwatch uses a cached version in memory.  This means that we are never able to
+	// unwatch without first watching.  If the manager crashes and comes back up again,
+	// there will be no possibility to unwatch existing groups.
+	m.lock.Lock()
+	config := m.currentConfig
+	m.lock.Unlock()
+
+	return m.doUnwatchGroups(config)
+}
+
+func (m *manager) doCommit() error {
+
+	// load the config
+	config := GlobalSpec{}
+
+	// load the latest version -- assumption here is that it's been persisted already.
+	err := m.snapshot.Load(&config)
+	if err != nil {
+		return err
+	}
+
+	log.Infoln("Committing.  Loaded snapshot. err=", err)
+	if err != nil {
+		return err
+	}
+
+	m.lock.Lock()
+	m.currentConfig = config // make a copy
+	m.lock.Unlock()
+
+	return m.doUpdateGroups(config)
+}
+
+func (m *manager) doUpdateGroups(config GlobalSpec) error {
+	return m.execPlugins(config,
+		func(plugin group.Plugin, spec group.Spec) error {
+
+			log.Infoln("UPDATE group", spec.ID, "with spec:", spec)
+			err := plugin.UpdateGroup(spec)
+
+			// TODO(chungers) -- yes this is clunky
+			if err != nil && strings.Contains(err.Error(), "not being watched") {
+
+				log.Infoln("UPDATE group", spec.ID, "changed to WATCH")
+				err = plugin.WatchGroup(spec)
+
+			}
+
+			if err != nil {
+				log.Warningln("Error updating/watch group:", spec.ID, "Err=", err)
+			}
+			return err
+		})
+}
+
+func (m *manager) doWatchGroups(config GlobalSpec) error {
+	log.Infoln("Start watching groups")
+	return m.execPlugins(config,
+		func(plugin group.Plugin, spec group.Spec) error {
+
+			log.Infoln("WATCH group", spec.ID, "with spec:", spec)
+			err := plugin.WatchGroup(spec)
+
+			// TODO(chungers) -- yes this is clunky
+			if err != nil && strings.Contains(err.Error(), "Already watching") {
+
+				log.Infoln("Already WATCHING", spec.ID, "no action")
+				return nil
+			}
+
+			if err != nil {
+				log.Warningln("Error watching group:", spec.ID, "Err=", err)
```

Comment:
Thinking more on this, while this condition does indicate a state mismatch that the user must reconcile, making this fatal could result in a state that's difficult to reconcile.  This is because the next attempt to restart this daemon would try to watch groups, and likely fail again indefinitely.